### ISSUE TITLE
pam_unix fails when calling from cron which runs in non-root Docker image

### ISSUE THREAD

**Issue by ViliusS** [linux-pam/linux-pam#788]
I'm trying to run cron ([cronie](https://github.com/cronie-crond/cronie) specifically) in rootless RHEL Docker image. Under RHEL cronie uses PAM. I have configured crond binary with `cap_setuid` and `cap_setgid` capabilities. This allows to run cronie from standard user account without issues (though binary setuids to root user after that).

Unfortunately, when cron tries to execute crontjobs of other users, it fails with `Authentication service cannot retrieve authentication info` error:
```
sh-5.1$ crond -x proc
debug flags enabled: proc
[275] cron started
log_it: (CRON 275) STARTUP (1.5.7)
log_it: (CRON 275) INFO (Syslog will be used instead of sendmail.)
log_it: (CRON 275) INFO (RANDOM_DELAY will be scaled with factor 25% if used.)
log_it: (CRON 275) INFO (running with inotify support)
log_it: (CRON 275) INFO (Can't create lock for reboot jobs.): Permission denied
[275] do_command(your_command_to_schedule, (default,1001,0))
[275] main process returning to work
log_it: (default 277) PAM ERROR (Authentication service cannot retrieve authentication info)
log_it: (default 277) FAILED to authorize user with PAM (Authentication service cannot retrieve authentication info)
[275] sigchld...pid #277 died, stat=1
[275] sigchld...no children
```
At first I thought that it's something in cronie permissions, but it runs as root:
```
sh-5.1$ ps aux
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
default      1  0.0  0.4  22088 14928 ?        Ss   11:57   0:00 httpd -D FOREGROUND
default     36  0.0  0.3 252720 11568 ?        Ss   11:57   0:00 php-fpm: master process (/etc/php-fpm.conf)
default     37  0.0  0.2 252720  8124 ?        S    11:57   0:00 php-fpm: pool www
default     38  0.0  0.2 252720  8124 ?        S    11:57   0:00 php-fpm: pool www
default     39  0.0  0.2 252720  8124 ?        S    11:57   0:00 php-fpm: pool www
default     40  0.0  0.2 252720  8124 ?        S    11:57   0:00 php-fpm: pool www
default     41  0.0  0.2 252720  8124 ?        S    11:57   0:00 php-fpm: pool www
root        43  0.0  0.0   5616  2708 ?        Ss   11:57   0:00 crond
default     48  0.0  0.0   5068  2432 ?        S    11:57   0:00 /usr/bin/coreutils --coreutils-prog-shebang=cat /usr/bin/cat
default     49  0.0  0.0   5068  2432 ?        S    11:57   0:00 /usr/bin/coreutils --coreutils-prog-shebang=cat /usr/bin/cat
default     50  0.0  0.0   5068  2560 ?        S    11:57   0:00 /usr/bin/coreutils --coreutils-prog-shebang=cat /usr/bin/cat
default     51  0.0  0.0   5068  2432 ?        S    11:57   0:00 /usr/bin/coreutils --coreutils-prog-shebang=cat /usr/bin/cat
default     52  0.0  0.1  21744  4456 ?        S    11:57   0:00 httpd -D FOREGROUND
default     53  0.0  0.2 2521428 8184 ?        Sl   11:57   0:00 httpd -D FOREGROUND
default     56  0.0  0.2 2324748 7800 ?        Sl   11:57   0:00 httpd -D FOREGROUND
default     82  0.0  0.2 2324748 7672 ?        Sl   11:57   0:00 httpd -D FOREGROUND
default    267  0.0  0.1   4452  3584 pts/0    Ss   11:57   0:00 /bin/sh
default    313  0.0  0.1   7148  3072 pts/0    R+   12:31   0:00 ps aux
```

I tracked it down to pam_unix.so module in /etc/pam.d/system-auth (included via /etc/pam.d/crond). I suspect it cannot access /etc/shadow file. I see that there were some fixes for sudo https://github.com/linux-pam/linux-pam/blob/cec36a8cd2c69cc87eacc21da471334fbef128ee/modules/pam_unix/pam_unix_acct.c#L113-L123 , but not sure why it doesn't work for cronie.

Shadow file is nothing extra ordinary:
```
bash-5.1# more /etc/shadow
root:!locked::0:99999:7:::
bin:*:19347:0:99999:7:::
daemon:*:19347:0:99999:7:::
adm:*:19347:0:99999:7:::
lp:*:19347:0:99999:7:::
sync:*:19347:0:99999:7:::
shutdown:*:19347:0:99999:7:::
halt:*:19347:0:99999:7:::
mail:*:19347:0:99999:7:::
operator:*:19347:0:99999:7:::
games:*:19347:0:99999:7:::
ftp:*:19347:0:99999:7:::
nobody:*:19347:0:99999:7:::
systemd-coredump:!!:19782::::::
dbus:!!:19782::::::
tss:!!:19782::::::
systemd-oom:!*:19782::::::
default:!!:19782::::::
apache:!!:19821::::::
nginx:!!:19821::::::
```

For full reproducible case build the following Dockerfile:
```
FROM registry.access.redhat.com/ubi9/php-81

USER root
RUN dnf install -y cronie && \
    dnf clean all && \
    chown default:root /var/run && setcap "cap_setuid=ep cap_setgid=ep" /usr/sbin/crond

USER default
COPY --chown=default:root php-pre-start/run-cron.sh ./php-pre-start/run-cron.sh

RUN (crontab -l; echo "* * * * * your_command_to_schedule") | crontab -

CMD /usr/libexec/s2i/run
```
php-pre-start/run-cron.sh contents:
```
#!/bin/bash

crond
```

The workaround is to add `broken_shadow` to pam_unix:
```
FROM registry.access.redhat.com/ubi9/php-81

USER root
RUN dnf install -y cronie && \
    dnf clean all && \
    chown default:root /var/run && setcap "cap_setuid=ep cap_setgid=ep" /usr/sbin/crond && \
    sed -i 's/\(account     required      pam_unix.so\)/\1 broken_shadow/g' /etc/pam.d/system-auth

# Reset to default application user
USER default
COPY --chown=default:root php-pre-start/run-cron.sh ./php-pre-start/run-cron.sh

RUN (crontab -l; echo "* * * * * your_command_to_schedule") | crontab -

CMD /usr/libexec/s2i/run
```

**Comment by cfhammill** [comment]
I ran into what I think is the same issue with nix provisioned docker containers using pam 1.6.1. 

The container was run as root; with pam and users set up roughly like https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/docker/default.nix#L206, but with an additional service user. I then couldn't `sudo -E -u service` with the error
```
sudo: PAM account management error: Permission denied
sudo: a password is required
```
worked around by adding broken_shadow as suggested by @ViliusS

**Comment by ViliusS** [comment]
Sorry about the confusion, let me rephrase. The /etc/shadow file itself is not modified. It doesn't have any additional permissions which are not there on standard Linux OS, i.e. its set to `chmod 0000`. However, as with any other file, this doesn't prevent a user with root privileged to be able to read it. You can test it by doing ordinary `cat /etc/shadow` from `root` account. This also doesn't prevent pam to read it if pam_unix module is called from service running under root account, let's say SSHD or cron daemon.

The issue starts when someone tries to call pam_unix module from a program with root privileged received through setuid. To test this, create any executable program which setuids to 0 and reads /etc/shadow directly. It will read it successfully. Now call cron executable which also setuids to 0 and uses pam_unix to read shadow file indirectly. It will fail.

**Comment by ViliusS** [comment]
Thank you, I will check your theory on SELinux part when I have time. I don't believe DAC override capability is at play here however. Unless I'm misremembering I don't run SELinux or AppArmor on container nodes. Also, ubi images do not include SELinux AFAIR, so any service under root permissions should be able to read any file inside a container.

In any case, even if dac_override helps, this doesn't explain why running sshd or cron directly as root works, but running from setuid 0 (which is essentially the same root just limited to the process running) doesn't work.

I have also read all discussion in linked moby issue. It starts to look to me that this is indeed a bug in PAM.
